python/ray/serve/_private/metrics_utils.py

```diff
@@ -307,71 +307,179 @@ def aggregate_avg(
+
+        for point in series[1:]:
+            t, v = point.timestamp, point.value
+            dv = v - prev_v
+            if dv != 0.0:
+                events[t] += dv
+            if ttl is not None:
+                # cancel previous expiry by adding back, then schedule new expiry
+                # (net effect: move expiry to t+ttl with the then-current value)
+                events[prev_t + ttl] += +prev_v  # cancel old expiry
+                events[t + ttl] += -v  # schedule new expiry
+                last_time[rid], last_val[rid] = t, v
+            prev_t, prev_v = t, v
+
+    # Sweep events (coalesce identical timestamps)
+    timeline = sorted(events.items())  # [(t, delta_sum_at_t), ...]
     merged: List[TimeStampedValue] = []
-    for w in windows:
-        v = b1.get(w, 0.0) + b2.get(w, 0.0)
-        ts_start = start + w * window_s
-        merged.append(TimeStampedValue(timestamp=ts_start, value=v))
+    current_sum = 0.0
+
+    for t, delta in timeline:
+        if delta == 0.0:
+            continue
+        current_sum += delta
+        # Record instantaneous total right after applying deltas at t
+        if merged and merged[-1].timestamp == t:
+            merged[-1] = TimeStampedValue(
+                t, current_sum
+            )  # coalesce exact same timestamp
+        else:
+            merged.append(TimeStampedValue(t, current_sum))
+
     return merged
 
 
 def merge_timeseries_dicts(
     *timeseries_dicts: DefaultDict[Hashable, List[TimeStampedValue]],
-    window_s: float,
+    ttl: Optional[float] = None,
 ) -> DefaultDict[Hashable, List[TimeStampedValue]]:
     """
-    Merge multiple time-series dictionaries, typically contained within
-    InMemoryMetricsStore().data. For the same key across stores, time series
-    are merged with a windowed sum, where each series keeps only its latest
-    value per window before summing.
+    Merge multiple time-series dictionaries using instantaneous merge approach.
     """
     merged: DefaultDict[Hashable, List[TimeStampedValue]] = defaultdict(list)
-    for timeseries_dict in timeseries_dicts:
-        for key, ts in timeseries_dict.items():
-            if key in merged:
-                merged[key] = _merge_two_timeseries(merged[key], ts, window_s)
-            else:
-                # Window the data, even if the key is unique.
-                merged[key] = _merge_two_timeseries(ts, [], window_s)
+
+    # Collect all unique keys
+    all_keys = set()
+    for ts_dict in timeseries_dicts:
+        all_keys.update(ts_dict.keys())
+
+    # Merge each key separately
+    for key in all_keys:
+        # Collect all series for this key across dictionaries
+        replicas_series = []
+        for ts_dict in timeseries_dicts:
+            if key in ts_dict:
+                replicas_series.append(ts_dict[key])
+
+        # Merge using instantaneous approach
+        merged[key] = merge_instantaneous_total(replicas_series, ttl=ttl)
+
     return merged
```

Comment:
I think you can do this in fewer passes with a default dict:

```suggestion
    metrics = defaultdict(list)
    for ts_dict in timeseries_dicts:
        for key, values in ts_dict.items();
            metrics[key].append(values)
            
    returned {k: merge_instantaneous_total(v, ttl=ttl) for k, v in metrics.items()}
```